manager/orchestrator/constraint_enforcer.go

```diff
@@ -0,0 +1,137 @@
+package orchestrator
+
+import (
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/constraint"
+	"github.com/docker/swarmkit/manager/state"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+// ConstraintEnforcer watches for updates to nodes and shuts down tasks that no
+// longer satisfy scheduling constraints or resource limits.
+type ConstraintEnforcer struct {
+	store       *store.MemoryStore
+	watcher     chan events.Event
+	cancelWatch func()
+	stopChan    chan struct{}
+	doneChan    chan struct{}
+}
+
+// NewConstraintEnforcer creates a new ConstraintEnforcer.
+func NewConstraintEnforcer(store *store.MemoryStore) *ConstraintEnforcer {
+	watcher, cancel := state.Watch(store.WatchQueue(), state.EventUpdateNode{})
+
+	return &ConstraintEnforcer{
+		store:       store,
+		watcher:     watcher,
+		cancelWatch: cancel,
+		stopChan:    make(chan struct{}),
+		doneChan:    make(chan struct{}),
+	}
+}
+
+// Run is the ConstraintEnforcer's main loop.
+func (ce *ConstraintEnforcer) Run() {
+	defer close(ce.doneChan)
+
+	for {
+		select {
+		case event := <-ce.watcher:
+			node := event.(state.EventUpdateNode).Node
```

Comment:
global orchestrator is also listening on EventUpdateNode. Can you check if there is race condition?